Refactoring: Split campaign types into classes

brayvasq opened the pull request:
## :loudspeaker: Change type
- [x] Refactoring

## :scroll: Description
<!--- Describe your changes in detail -->
- [x] Add classes for each type of Campaingn (sms, web_push, push, email).
- [x] Add 'bulk_create_campaign' method.

## :green_heart: Testing
Specs

brayvasq committed:
change: Split campaign types into classes

2xG commented on lib/clever_tap.rb:
```diff
@@ -3,6 +3,10 @@
 require 'clever_tap/config'
 require 'clever_tap/client'
 require 'clever_tap/campaign'
+require 'clever_tap/sms_campaign'
```
**suggestion**: `clever_tap/campaign/sms` etc

2xG replied:
**suggestion**:  `CleverTap::EmailCampaign` -> `CleverTap::Campaign::Email`

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -104,12 +103,6 @@ def campaign_id_hash
 
     def content_hash
```
**suggestion**:
```ruby
raise NotImplementedError
```
aqui y en otros lugares agregar este tipo de errores que no permiten usar la clase `Campaign` directamente.

2xG commented on lib/clever_tap/campaign_creator.rb:
```diff
@@ -16,9 +16,9 @@ class CampaignCreator
     attr_reader :record, :type
 
-    def initialize(record, type: :sms)
+    def initialize(record)
```
**suggestion**: `record` -> `campaign`

**thought**: campaign tiene mas sentido que record, record es mas como entidad de DB

2xG replied:
veo que en `Uploader.initialize` manejan `records`, pero los perfiles y eventos parecen records =)

2xG commented on lib/clever_tap/campaign_creator.rb:
```diff
@@ -29,6 +29,19 @@ def call(client)
 
     private
 
+    def campaign_type(record)
```
**suggestion**: `campaign_type(record)` -> `type_of(campaign)`

2xG commented on lib/clever_tap/campaign_creator.rb:
```diff
@@ -29,6 +29,19 @@ def call(client)
+        TYPE_PUSH
+      when CleverTap::EmailCampaign
+        TYPE_EMAIL
+      end
```
**chore**: agregar `else`

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -147,20 +140,7 @@ def mutable_content_hash
-
-      { PLATFORM_SPECIFIC => @platform_specific }
-    end
+    def platform_specific_hash; end
```
**question**: tiene sentido tenerlo aqui?

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -147,20 +140,7 @@ def mutable_content_hash
-    end
+    def platform_specific_hash; end
 
     def content_platform_specific
```
**question**: que proposito tiene?

brayvasq replied:
Se puede quitar. Realmente se está validando ese hash en las clases WebPush y Push, ya que al final eso va en el content de esas dos clases.

Lo había dejado ahí por que en la documentación aparecía como un parámetro independiente y no dentro de content.

brayvasq committed:
add: bulk_create_campaign
  add: Campaign 'receivers_limit_exceeded?' method to know if the limit was exceeded
  add: Client 'obtain_receivers' method, to split receivers in campaigns which do not exceed the limit
  add: Client 'bulk_create_campaign' method, allowing to send campaigns with more than 1000 receivers
  change: check if receivers limit was exeeded in the 'create_campaign' method

2xG commented on lib/clever_tap.rb:
```diff
@@ -68,8 +72,8 @@ def upload_profile(profile, **options)
 
-  def create_campaign(campaign, type: :sms)
-    response = CampaignCreator.new(campaign, type: type).call(client)
+  def create_campaign(campaign)
```
**thought**: parece que los métodos de la clase `CleverTap` como tal están planeados para morir

Desde README.md
> Available in v0.2.0 but will be depricated as of v1.0.0 `.upload_event` accepts as a first argument an object responding to `#to_h` and `#[]` and a second parameter keyword argument `name: <name>`.

2xG commented on lib/clever_tap/client.rb:
```diff
@@ -68,12 +68,73 @@ def upload(records, dry_run: 0)
 
     private
 
+    def obtain_receivers(campaign)
```
**suggestion**: simplificar la logica: transformar el hash `campaign.to` a un array de pares `[id_type, id]`:

```ruby
identitiy_pairs = campaign.to.map { |type, list| list.map { |id| [type, id] } }
                          .flatten(1)
pp identitiy_pairs
```

en esta etapa tenemos algo asi en la salida

``` ruby
[
    ["FBID", "102029292929388"],
    ["FBID", "114342342453463"],
    ["GPID", "1928288389299292"],
    ["Email", "john@doe.com"],
    ["Email", "jane@doe.com"],
    ["Identity", "JohnDoe"],
    ["objectId", "_asdnkansdjknaskdjnasjkndja"],
    ["objectId", "-adffajjdfoaiaefiohnefwprjf"]
]

```

después cortarlo en pedazos de 1000 ids y volver a armar los hashes de estos pedazos:

```ruby
identitiy_pairs.each_slice(Campaign::MAX_USERS_PER_CAMPAIGN) do |slice|
  chunked_to =  slice.group_by(&:first)
                     .map { |k, v| [k, v.map(&:last)] }
                     .to_h
  pp chunked_to
end
```

aquí dentro del bloque `each_slice` podemos crear nuevas campañas con `to = chunked_to` y enviarlas.
en el ejemplo de arriba, asumiendo `Campaign::MAX_USERS_PER_CAMPAIGN = 2` en la consola tengo:

```ruby
{"FBID"=>["102029292929388", "114342342453463"]}
{"GPID"=>["1928288389299292"], "Email"=>["john@doe.com"]}
{"Email"=>["jane@doe.com"], "Identity"=>["JohnDoe"]}
{"objectId"=>["_asdnkansdjknaskdjnasjkndja", "-adffajjdfoaiaefiohnefwprjf"]}
```

2xG replied:
me llego la idea que puedes poner `yield chunked_to` dentro del bloque `each_slice` y dentro del método `create_campaign` pasar el bloque (y cambiar el nombre del metodo `obtain_receivers` -> `receivers_chunks`
```ruby
def create_campaign(campaign)
  responses = []

  receivers_chunks(campaign) do |receivers|
    campaign.to = receivers
    responses << CampaignCreator.new(campaign).call(self)
  end

  responses
end
```

2xG commented on lib/clever_tap/client.rb:
```diff
@@ -68,12 +68,73 @@ def upload(records, dry_run: 0)
-    def create_campaign(campaign, type: :sms)
-      CampaignCreator.new(campaign, type: type).call(self)
+    def create_campaign(campaign)
+      if campaign.receivers_limit_exceeded?
```
**suggestion**: moverlo al metodo `obtain_receivers`

```ruby
def obtain_receivers(campaign)
  return [campaign.to] unless campaign.receivers_limit_exceeded?
...
end
```
y dejar un solo método de crear campañas `create_campaign` pero de manera `bulk_create_campaign`. es decir si no son mas de 1000 identificadores devolvemos la lista de manera unificada: array con un solo elemento

2xG commented on lib/clever_tap/email_campaign.rb:
```diff
@@ -0,0 +1,28 @@
+class CleverTap
+  class EmailCampaign < Campaign
```
**suggestion**: `class Campaign::Email`

2xG commented on lib/clever_tap/push_campaign.rb:
```diff
@@ -0,0 +1,50 @@
+class CleverTap
+  class PushCampaign < Campaign
```
**suggestion**: `class Campaign::Push`

2xG commented on lib/clever_tap/push_campaign.rb:
```diff
@@ -0,0 +1,50 @@
+    def platform_specific_hash
+      return {} unless @platform_specific
+
+      android = @platform_specific[:android] || @platform_specific['android']
```
**suggestion**: creo que puedes usar `symbolize_keys` o `symbolize_keys!` dentro del `initialize` para simplificar estos condiciones aqui y en otros lgares

brayvasq replied:
@2xG Creo que ese es un método de rails, no de ruby en sí.

2xG commented on lib/clever_tap/sms_campaign.rb:
```diff
@@ -0,0 +1,25 @@
+class CleverTap
+  class SmsCampaign < Campaign
```
**suggestion**: `class Campaign::Sms`

2xG commented on lib/clever_tap/web_push_campaign.rb:
```diff
@@ -0,0 +1,41 @@
+class CleverTap
+  class WebPushCampaign < Campaign
```
**suggestion**: `class Campaign::WebPush`

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -87,7 +88,7 @@ def receivers_hash
       raise InvalidIdentityTypeError unless allowed?(@to.keys)
       raise NoReceiversError if @to.values.all?(&:empty?)
-      raise ReceiversLimitExceededError if @to.values.map(&:size).reduce(&:+) > MAX_USERS_PER_CAMPAIGN
+      raise ReceiversLimitExceededError if receivers_limit_exceeded?
```
**praise**: :+1: se ve mas limpio

brayvasq committed:
change: move specific campaigns model to the Campaign scope

brayvasq committed:
change: refactoring the bulk_create_campaign
  change: campaign variable names from 'record' to 'campaign'
  add: empty_receivers? method to validate if there are no receivers
  change: rename bulk_create_campaign to receivers_chunks
  fix: use yield_control expectations

2xG commented on lib/clever_tap/client.rb:
```diff
@@ -69,20 +69,12 @@ def upload(records, dry_run: 0)
-        CampaignCreator.new(campaign).call(self)
-      end
-    end
+      return [CampaignCreator.new(campaign).call(self)] unless campaign.receivers_limit_exceeded?
```
perdon creo que no explique bien.
la idea fue de meter esto en `receivers_chunks` para evitar el uso doble de `CampaignCreator.new(campaign).call(self)`, pero como se cambio la idea como usar este método eso debería ser así
```ruby
def receivers_chunks(campaign)
  chunked_to = campaign.to unless campaign.receivers_limit_exceeded?
  ...
end
```
me hago entender?

brayvasq replied:
No, no entiendo. En ese caso se necesitaría otro yield?

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -84,10 +85,9 @@ def to_h
 
     def receivers_hash
-      raise NoReceiversError if @to.to_h.empty?
+      raise NoReceiversError if empty_receivers?
```
**praise**: :+1:

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -84,10 +85,9 @@ def to_h
       raise InvalidIdentityTypeError unless allowed?(@to.keys)
-      raise NoReceiversError if @to.values.all?(&:empty?)
-      raise ReceiversLimitExceededError if @to.values.map(&:size).reduce(&:+) > MAX_USERS_PER_CAMPAIGN
+      raise ReceiversLimitExceededError if receivers_limit_exceeded?
```
**praise**: :+1:

brayvasq committed:
change: remove unnecesary guard condition in the create_campaign method

2xG commented on lib/clever_tap/client.rb:
```diff
@@ -68,12 +68,31 @@ def upload(records, dry_run: 0)
     private
 
+    def receivers_chunks(campaign)
+      identity_pairs = campaign.to.flat_map { |type, list| list.map { |id| [type, id] } }
```
**praise**: no sabia que existe `flat_map` :smiley:

brayvasq committed:
fix: remove duplicated shared example in 'spec/shared/clever_tap_client.rb'
